src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkRuleContext.java

```diff
@@ -1143,9 +1143,14 @@ public static ImmutableMap<Label, ImmutableCollection<Artifact>> makeLabelMap(
     ImmutableMap.Builder<Label, ImmutableCollection<Artifact>> builder = ImmutableMap.builder();
 
     for (TransitiveInfoCollection current : knownLabels) {
+      Label label = AliasProvider.getDependencyLabel(current);
+      FilesToRunProvider filesToRun = current.getProvider(FilesToRunProvider.class);
```

Comment:
This is tied to another issue I have seen many people hitting:

When a custom rule returns the `DefaultInfo` with `files` set but no `runfiles`, and then passing a target constructed with this custom rule to a test, the `ctx.expand_location` will expand, but the file won't be present during the test execution.

I don't think, though, the right answer is to just move away from "files to build" to "files to run" since the location expander could be used in actions at build time.

The answer is probably to look at the runfiles first and fall back to files.

Reply:
Addressed in 941c8f5298.

Reply:
I think you largely meant FilesToRun instead of runfiles in your post?

There's 3 sets of files with similar names:
FilesToRunProvider is equiv to DefaultInfo.files_to_run
RunfilesProvider is equiv to DefaultInfo.{default,data}_runfiles
FilesProvider ("files to build") is equiv to DefaultInfo.files

Ahhh, something I just noticed is the LocationExpander logic, for the deps/data/tools attributes, falls back to FilesToRun.getFilesToRun (I misread it as FilesProvider.getFilestoBuild). That seems weird. This gives me a very historical-code vibe. You know, I wouldn't be surprised if we look around and find something like filegroup or genrule (which are often put in data) are putting their files in FilesToRunProvider.filesToBuild, but don't set an executable (because they aren't) just so they could get properly expanded in $(location) expressions.

Yeah, I guess just copy that behavior: executable, then files to run, then files to build.

Reply:
Yes, I was mixing runfiles and files to run.